metrics/sinks/file/FileSink.cpp

```diff
@@ -0,0 +1,94 @@
+/*##############################################################################
+    HPCC SYSTEMS software Copyright (C) 2021 HPCC Systems®.
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "FileSink.hpp"
+#include <cstdio>
+#include <thread>
+
+using namespace hpccMetrics;
+
+extern "C" IMetricSink* getSinkInstance(const char *name, const IPropertyTree *pSettingsTree)
+{
+    std::string sinkName(name);
+    IMetricSink *pSink = new FileMetricSink(name, pSettingsTree);
+    return pSink;
+}
+
+
+FileMetricSink::FileMetricSink(const char *name, const IPropertyTree *pSettingsTree) :
+        MetricSink(name, "file"),
+        periodSeconds{std::chrono::seconds(60)}
+{
+    if (pSettingsTree->hasProp("@period"))
+    {
+        unsigned seconds = pSettingsTree->getPropInt("@period");
+        periodSeconds = std::chrono::seconds(seconds);
+    }
+
+    pSettingsTree->getProp("@filename", fileName);
+    bool clearFile = pSettingsTree->getPropBool("@clear", false);
+
+    //
+    // Clear the file if indicated.
+    if (clearFile)
+    {
+        auto handle = fopen(fileName.str(), "w");
+        if (handle != nullptr)
+        {
+            fclose(handle);
+        }
+    }
+}
+
+
+void FileMetricSink::startCollection(MetricsReporter *_pReporter)
+{
+    pReporter = _pReporter;
+    collectThread = std::thread(&FileMetricSink::collectionThread, this);
+}
+
+
+void FileMetricSink::collectionThread() const
+{
+    while (!stopCollectionFlag)
+    {
+        std::this_thread::sleep_for(periodSeconds);
+        auto reportMetrics = pReporter->getReportMetrics(name);
+        MeasurementVector measurements;
+        for (auto &pMetric: reportMetrics)
+        {
+            pMetric->getReportValues(measurements);
+        }
+        writeMeasurementsToFile(measurements);
+    }
+}
+
+
+void FileMetricSink::stopCollection()
+{
+    stopCollectionFlag = true;
+    collectThread.join();
+}
+
+
+void FileMetricSink::writeMeasurementsToFile(const MeasurementVector &values) const
+{
+    auto fh = fopen(fileName.str(), "a");
```

Comment:
minor: Should probably use jfile functions.
Is it correct to open and close the file each time?
Minor comments since this is unlikely to be used for anything other than debugging.